File: lecture11/exercise11-A/solution.cpp
// Exercise 11-A: Lotto

#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

// Constants

const int LOTTO_NUMBERS = 6;

// Functions

void display_combinations(vector<int> &s, vector<int> &d, size_t k) {
    // Base: have complete subset
    if (k == d.size()) {
    	if (s.size() == LOTTO_NUMBERS) {
	    cout << s[0];
	    for (size_t i = 1; i < s.size(); i++) {
		cout << " " << s[i]; 
	    } 
	    cout << endl;
	}
    	return;
    }
    
    // Recurse: with current
    s.push_back(d[k]);
    display_combinations(s, d, k + 1);
    s.pop_back(); // Reset subset

    // Recurse: skip current
    display_combinations(s, d, k + 1);
}

// Main execution

int main(int argc, char *argv[]) {
    string line;
    for (size_t line_count = 0; getline(cin, line); line_count++) {
    	stringstream ss(line);
    	int k;
    	ss >> k;
    	if (!k) {
    	    break;
	}

	int n;
	vector<int> numbers;
	vector<int> combination;
	while (ss >> n) {
	    numbers.push_back(n);
	}
	
	sort(numbers.begin(), numbers.end());

	if (line_count) {
	    cout << endl;
	}

	display_combinations(combination, numbers, 0);
    }

    return 0;
}

File: lecture11/exercise11-A/solution.py
#!/usr/bin/env python3

# Exercise 11-A: Lotto

import itertools
import sys

# Constants

LOTTO_NUMBERS = 6

# Functions

def display_combinations(s):
    for combination in sorted(itertools.combinations(s, LOTTO_NUMBERS)):
        print(' '.join(map(str, combination)))

def display_combinations(s, d, k):
    '''
    s:  Current combination
    d:  Numbers to choose from
    k:  Current element index
    '''
    if k == len(d):
        if len(s) == LOTTO_NUMBERS:
            print(' '.join(map(str, s)))
        return

    display_combinations(s + [d[k]], d, k + 1)
    display_combinations(s         , d, k + 1)

# Main execution

def main():
    for line, numbers in enumerate(map(str.split, sys.stdin)):
        if len(numbers) <= 1:
            break
        
        if line:
            print('')

        display_combinations([], list(map(int, numbers[1:])), 0)

if __name__ == '__main__':
    main()
